### ISSUE TITLE
libbpf: add legacy kprobe attach support

### ISSUE THREAD

**Issue by anakryiko** [libbpf/libbpf#317]
libbpf is currently only supporting attaching to kprobe/kretprobe using auto-detachable FD-based API, which is available since 4.16 (I think). But the need to support kprobe/kretprobe attachment on older kernels came up time and time again.

libbpf has bpf_link abstraction, which provides explicit destructor (bpf_link__destroy()), which is a perfect place to do manual clean up. It's not as safe as newer API and won't auto-clean up BPF program from the system on crash, but it gives often requested feature to the users on old kernels. And BCC supports this legacy kprobe anyway, so it would be good to have a feature parity here.

*N.B.* There is a previous attempt to implement this which hasn't received any update for a while: https://lore.kernel.org/bpf/20210318062520.3838605-1-rafaeldtinoco@ubuntu.com/

Example code below implements most of the functionality. It needs to be integrated properly into libbpf with feature probing (to detect which of the two APIs to use, depending on kernel feature support).

```c
static void bump_memlock_limit() {
  rlimit rlim = {
      .rlim_cur = 512UL << 20,
      .rlim_max = 512UL << 20,
  };
  int err = setrlimit(RLIMIT_MEMLOCK, &rlim);
  if (err) {
    fprintf(stderr, "Failed to setrlimit(): %d\n", -errno);
    exit(1);
  }
}

static void handle_shutdown(int /*signal*/) {
  shutdown = true;

  signal(SIGINT, SIG_DFL);
  signal(SIGTERM, SIG_DFL);
}

int poke_kprobe_events(bool add, const char* name, bool ret) {
  char buf[256];
  int fd, err;

  fd = open("/sys/kernel/debug/tracing/kprobe_events", O_WRONLY | O_APPEND, 0);
  if (fd < 0) {
    err = -errno;
    fprintf(stderr, "failed to open kprobe_events file: %d\n", err);
    return err;
  }

  if (add)
    snprintf(buf, sizeof(buf), "%c:kprobes/%s %s", ret ? 'r' : 'p', name, name);
  else
    snprintf(buf, sizeof(buf), "-:kprobes/%s", name);

  err = write(fd, buf, strlen(buf));
  if (err < 0) {
    err = -errno;
    fprintf(
        stderr,
        "failed to %s kprobe '%s': %d\n",
        add ? "add" : "remove",
        buf,
        err);
  }
  close(fd);
  return err >= 0 ? 0 : err;
}

int add_kprobe_event(const char* func_name, bool is_kretprobe) {
  return poke_kprobe_events(true /*add*/, func_name, is_kretprobe);
}

int remove_kprobe_event(const char* func_name, bool is_kretprobe) {
  return poke_kprobe_events(false /*remove*/, func_name, is_kretprobe);
}

struct bpf_link* attach_kprobe_legacy(
    struct bpf_program* prog,
    const char* func_name,
    bool is_kretprobe) {
  char fname[256], buf[256];
  struct perf_event_attr attr;
  struct bpf_link* link;
  int fd = -1, err, id;
  FILE* f = NULL;

  err = add_kprobe_event(func_name, is_kretprobe);
  if (err) {
    fprintf(stderr, "failed to create kprobe event: %d\n", err);
    return NULL;
  }

  snprintf(
      fname,
      sizeof(fname),
      "/sys/kernel/debug/tracing/events/kprobes/%s/id",
      func_name);
  f = fopen(fname, "r");
  if (!f) {
    fprintf(stderr, "failed to open kprobe id file '%s': %d\n", fname, -errno);
    goto err_out;
  }

  if (fscanf(f, "%d\n", &id) != 1) {
    fprintf(stderr, "failed to read kprobe id from '%s': %d\n", fname, -errno);
    goto err_out;
  }

  fclose(f);
  f = NULL;

  memset(&attr, 0, sizeof(attr));
  attr.size = sizeof(attr);
  attr.config = id;
  attr.type = PERF_TYPE_TRACEPOINT;
  attr.sample_period = 1;
  attr.wakeup_events = 1;

  fd = syscall(__NR_perf_event_open, &attr, -1, 0, -1, PERF_FLAG_FD_CLOEXEC);
  if (fd < 0) {
    fprintf(
        stderr,
        "failed to create perf event for kprobe ID %d: %d\n",
        id,
        -errno);
    goto err_out;
  }

  link = bpf_program__attach_perf_event(prog, fd);
  err = libbpf_get_error(link);
  if (err) {
    fprintf(stderr, "failed to attach to perf event FD %d: %d\n", fd, err);
    goto err_out;
  }

  return link;

err_out:
  if (f)
    fclose(f);
  if (fd >= 0)
    close(fd);
  remove_kprobe_event(func_name, is_kretprobe);
  return NULL;
}

int main(int argc, char** argv) {
  if (getuid() != 0) {
    fprintf(stderr, "should run as root!\n");
    return 1;
  }

  bump_memlock_limit();

  libbpf_set_print(libbpf_print_fn);

  printf("Opening and loading BPF skeleton...\n");
  struct kprobe_bpf* skel = kprobe_bpf__open_and_load();
  if (!skel) {
    fprintf(stderr, "Failed to open and load BPF skeleton\n");
    return 1;
  }

  __u32 zero = 0, my_pid = getpid();
  int err = bpf_map_update_elem(
      bpf_map__fd(skel->maps.my_pid_map), &zero, &my_pid, BPF_ANY);
  if (err) {
    fprintf(stderr, "Failed to set my PID param: %d\n", my_pid);
    err = 1;
    goto out;
  }

  fprintf(stderr, "Setting up interrupt handlers...\n");
  signal(SIGINT, &handle_shutdown);
  signal(SIGTERM, &handle_shutdown);

  fprintf(stderr, "Attaching kprobe...\n");
  skel->links.handler = attach_kprobe_legacy(
      skel->progs.handler, "hrtimer_start_range_ns", false /* is_kretprobe */);
  if (!skel->links.handler) {
    fprintf(stderr, "Failed to attach kprobe using legacy debugfs API!\n");
    scanf("%d", &err);
    err = 1;
    goto out;
  }

  fprintf(stderr, "Press Ctrl-C to stop...");
  while (!shutdown) {
    usleep(100000); /* each one triggers our kprobe */
  }

out:
  kprobe_bpf__destroy(skel);
  remove_kprobe_event("hrtimer_start_range_ns", false /* is_kretprobe */);
  return err;
}
```

BPF program for completeness (it doesn't depend on the legacy vs modern kprobe interface usage):

```c
#include "vmlinux.h>"

#include "bpf/bpf_helpers.h"
#include "bpf/bpf_tracing.h"

struct {
  __uint(type, BPF_MAP_TYPE_ARRAY);
  __uint(max_entries, 1);
  __type(key, __u32);
  __type(value, __u32);
} my_pid_map SEC(".maps");

SEC("kprobe/hrtimer_start_range_ns")
int BPF_KPROBE(
    handler,
    struct hrtimer* timer,
    ktime_t tim,
    u64 delta_ns,
    const enum hrtimer_mode mode) {
  u64 pid_tgid = bpf_get_current_pid_tgid();
  __u32 pid = pid_tgid >> 32;
  __u32 *my_pid, zero = 0;

  my_pid = bpf_map_lookup_elem(&my_pid_map, &zero);
  if (!my_pid || *my_pid != pid)
    return 0;

  /* to see this output:
   * $ sudo cat /sys/kernel/debug/tracing/trace_pipe
   */
  bpf_printk(
      "PID %d: timeout = %lld ns, delta = %lld ns\n", pid, tim, delta_ns);
  return 0;
}

char LICENSE[] SEC("license") = "GPL";
```

**Comment by rafaeldtinoco** [comment]
Hello Andrii,

Sorry for not replying to the thread any longer, I'll get back to this soon if nobody else does. By soon I mean ~1 week most. I have to deal with the kprobe disabling thing discussed in the last e-mail we shared for this.

Sorry again for the bad reply time.

**Comment by anakryiko** [comment]
Ok, thanks for update. I wasn't sure if you have an intention of finishing that feature.

**Comment by rafaeldtinoco** [comment]
https://github.com/rafaeldtinoco/libbpf/commit/e5be2dcbf4aedcc0a8d80b6eb0c74338bc93a243

I think that this approach here:

https://github.com/rafaeldtinoco/libbpf/commit/e5be2dcbf4aedcc0a8d80b6eb0c74338bc93a243#diff-ac82747bd450e8e5e0805a99d6c24afa9fcbfcddffbce5aa3288b2edabe17419L10302

is best as it only adds a single perf_event_open_probe_legacy() function to differentiate from what libbpf already has.

> I still have to fix the "remove_kprobe_event_legacy" part (workaround for being unable to remove kprobes from kprobe_events until after "bpf_object__detach_skeleton and bpf_object_close" are called.

> Remove other parts discussed [HERE](https://lore.kernel.org/bpf/CAEf4BzY09+RUNmVU4SQ5gDcxH4X_zCqh=rSGYhHwMseHNwZedw@mail.gmail.com/)

**Comment by rainkin1993** [comment]
@anakryiko 
> libbpf is currently only supporting attaching to kprobe/kretprobe using auto-detachable FD-based API, which is available since 4.16 (I think). 
I try to make clear the minimal kernel requirement of existing libbpf + kprobe.

1. Theoritically, the kernel support creating [k,u]probe with perf_event_open is 4.17-rc.
Here is the commit and the kernel release version: https://github.com/torvalds/linux/commit/65074d43fc77bcae32776724b7fa2696923c78e4

2. But actually existing libbpf + kprobe in kenel works in 4.17 and 4.19 while not working in 4.18.

- **Test code:**
**kernel code:**
```
#include <linux/bpf.h>
#include <bpf/bpf_helpers.h>
#include <bpf/bpf_tracing.h>
#include <bpf/bpf_core_read.h>
char LICENSE[] SEC("license") = "Dual BSD/GPL";
SEC("kprobe/vfs_write")
int BPF_KPROBE(vfs_write)
{
    return 0;
}
```
For the **user land code**, I just reuse the one in libbpf-bootstrap https://github.com/libbpf/libbpf-bootstrap/blob/master/examples/c/kprobe.c, which, I think, will print all debug information.

For the **libbpf version**, I also use the same as libbpf-bootstrap, that is https://github.com/libbpf/libbpf/tree/767d82caab7e54238f2fc6f40ab1e4af285f2abe

- **Test environment**: Ubuntu kernel 4.19.0-041900-generic

**Result: WORK.** The following is the output:
```
libbpf: loading object 'kprobe_bpf' from buffer
libbpf: elf: section(2) kprobe/vfs_write, size 16, link 0, flags 6, type=1
libbpf: sec 'kprobe/vfs_write': found program 'vfs_write' at insn offset 0 (0 bytes), code size 2 insns (16 bytes)
libbpf: elf: section(3) license, size 13, link 0, flags 3, type=1
libbpf: license of kprobe_bpf is Dual BSD/GPL
libbpf: elf: section(4) .BTF, size 371, link 0, flags 0, type=1
libbpf: elf: section(5) .BTF.ext, size 80, link 0, flags 0, type=1
libbpf: elf: section(6) .symtab, size 96, link 10, flags 0, type=2
libbpf: looking for externs among 4 symbols...
libbpf: collected 0 externs total
Successfully started! Please run `sudo cat /sys/kernel/debug/tracing/trace_pipe` to see output of the BPF programs.
```

**Configs:**
```
CONFIG_CGROUP_BPF=y
CONFIG_BPF=y
CONFIG_BPF_SYSCALL=y
CONFIG_BPF_JIT_ALWAYS_ON=y
CONFIG_IPV6_SEG6_BPF=y
CONFIG_NETFILTER_XT_MATCH_BPF=m
CONFIG_BPFILTER=y
CONFIG_BPFILTER_UMH=m
CONFIG_NET_CLS_BPF=m
CONFIG_NET_ACT_BPF=m
CONFIG_BPF_JIT=y
CONFIG_BPF_STREAM_PARSER=y
CONFIG_LWTUNNEL_BPF=y
CONFIG_HAVE_EBPF_JIT=y
CONFIG_BPF_EVENTS=y
CONFIG_BPF_KPROBE_OVERRIDE=y
CONFIG_KPROBES=y
CONFIG_KPROBES_ON_FTRACE=y
CONFIG_HAVE_KPROBES=y
CONFIG_HAVE_KPROBES_ON_FTRACE=y
CONFIG_KPROBE_EVENTS=y
# CONFIG_KPROBE_EVENTS_ON_NOTRACE is not set
CONFIG_BPF_KPROBE_OVERRIDE=y
```

- **Test environment**: Ubuntu kernel  4.18.0-18-generic

**Result: NOT WORK.** The following is the output:
```
libbpf: loading object 'kprobe_bpf' from buffer
libbpf: elf: section(2) kprobe/vfs_write, size 16, link 0, flags 6, type=1
libbpf: sec 'kprobe/vfs_write': found program 'vfs_write' at insn offset 0 (0 bytes), code size 2 insns (16 bytes)
libbpf: elf: section(3) license, size 13, link 0, flags 3, type=1
libbpf: license of kprobe_bpf is Dual BSD/GPL
libbpf: elf: section(4) .BTF, size 371, link 0, flags 0, type=1
libbpf: elf: section(5) .BTF.ext, size 80, link 0, flags 0, type=1
libbpf: elf: section(6) .symtab, size 96, link 10, flags 0, type=2
libbpf: looking for externs among 4 symbols...
libbpf: collected 0 externs total
libbpf: load bpf program failed: Invalid argument
libbpf: failed to load program 'vfs_write'
libbpf: failed to load object 'kprobe_bpf'
libbpf: failed to load BPF skeleton 'kprobe_bpf': -22
Failed to open BPF skeleton
```

**Configs:**
```
CONFIG_CGROUP_BPF=y
CONFIG_BPF=y
CONFIG_BPF_SYSCALL=y
CONFIG_BPF_JIT_ALWAYS_ON=y
CONFIG_IPV6_SEG6_BPF=y
CONFIG_NETFILTER_XT_MATCH_BPF=m
CONFIG_BPFILTER=y
CONFIG_BPFILTER_UMH=m
CONFIG_NET_CLS_BPF=m
CONFIG_NET_ACT_BPF=m
CONFIG_BPF_JIT=y
CONFIG_BPF_STREAM_PARSER=y
CONFIG_LWTUNNEL_BPF=y
CONFIG_HAVE_EBPF_JIT=y
CONFIG_BPF_EVENTS=y
CONFIG_BPF_KPROBE_OVERRIDE=y
CONFIG_TEST_BPF=m
CONFIG_KPROBES=y
CONFIG_KPROBES_ON_FTRACE=y
CONFIG_HAVE_KPROBES=y
CONFIG_HAVE_KPROBES_ON_FTRACE=y
CONFIG_KPROBE_EVENTS=y
CONFIG_BPF_KPROBE_OVERRIDE=y
# CONFIG_KPROBES_SANITY_TEST is not set
```

- **Test environment**: Ubuntu kernel 4.17.5-041705-generic

**Result: WORK.** The following is the output:
```
libbpf: loading object 'kprobe_bpf' from buffer
libbpf: elf: section(2) kprobe/vfs_write, size 16, link 0, flags 6, type=1
libbpf: sec 'kprobe/vfs_write': found program 'vfs_write' at insn offset 0 (0 bytes), code size 2 insns (16 bytes)
libbpf: elf: section(3) license, size 13, link 0, flags 3, type=1
libbpf: license of kprobe_bpf is Dual BSD/GPL
libbpf: elf: section(4) .BTF, size 371, link 0, flags 0, type=1
libbpf: elf: section(5) .BTF.ext, size 80, link 0, flags 0, type=1
libbpf: elf: section(6) .symtab, size 96, link 10, flags 0, type=2
libbpf: looking for externs among 4 symbols...
libbpf: collected 0 externs total
libbpf: Kernel doesn't support BTF, skipping uploading it.
Successfully started! Please run `sudo cat /sys/kernel/debug/tracing/trace_pipe` to see output of the BPF programs.
```

**Configs:**
```
CONFIG_CGROUP_BPF=y
CONFIG_BPF=y
CONFIG_BPF_SYSCALL=y
CONFIG_BPF_JIT_ALWAYS_ON=y
CONFIG_NETFILTER_XT_MATCH_BPF=m
CONFIG_NET_CLS_BPF=m
CONFIG_NET_ACT_BPF=m
CONFIG_BPF_JIT=y
CONFIG_BPF_STREAM_PARSER=y
CONFIG_LWTUNNEL_BPF=y
CONFIG_HAVE_EBPF_JIT=y
CONFIG_BPF_EVENTS=y
CONFIG_BPF_KPROBE_OVERRIDE=y
CONFIG_KPROBES=y
CONFIG_KPROBES_ON_FTRACE=y
CONFIG_HAVE_KPROBES=y
CONFIG_HAVE_KPROBES_ON_FTRACE=y
CONFIG_KPROBE_EVENTS=y
CONFIG_BPF_KPROBE_OVERRIDE=y
```

- **Test environment**: Ubuntu kernel 4.15.0-101-generic

**Result: NOT WORK.** The following is the output:
```
libbpf: loading object 'kprobe_bpf' from buffer
libbpf: elf: section(2) kprobe/vfs_write, size 16, link 0, flags 6, type=1
libbpf: sec 'kprobe/vfs_write': found program 'vfs_write' at insn offset 0 (0 bytes), code size 2 insns (16 bytes)
libbpf: elf: section(3) license, size 13, link 0, flags 3, type=1
libbpf: license of kprobe_bpf is Dual BSD/GPL
libbpf: elf: section(4) .BTF, size 371, link 0, flags 0, type=1
libbpf: elf: section(5) .BTF.ext, size 80, link 0, flags 0, type=1
libbpf: elf: section(6) .symtab, size 96, link 10, flags 0, type=2
libbpf: looking for externs among 4 symbols...
libbpf: collected 0 externs total
libbpf: Kernel doesn't support BTF, skipping uploading it.
libbpf: load bpf program failed: Invalid argument
libbpf: failed to load program 'vfs_write'
libbpf: failed to load object 'kprobe_bpf'
libbpf: failed to load BPF skeleton 'kprobe_bpf': -22
Failed to open BPF skeleton
```

**Configs**
```
CONFIG_CGROUP_BPF=y
CONFIG_BPF=y
CONFIG_BPF_SYSCALL=y
CONFIG_BPF_JIT_ALWAYS_ON=y
CONFIG_NETFILTER_XT_MATCH_BPF=m
CONFIG_NET_CLS_BPF=m
CONFIG_NET_ACT_BPF=m
CONFIG_BPF_JIT=y
CONFIG_BPF_STREAM_PARSER=y
CONFIG_LWTUNNEL_BPF=y
CONFIG_HAVE_EBPF_JIT=y
CONFIG_BPF_EVENTS=y
CONFIG_TEST_BPF=m
CONFIG_KPROBES=y
CONFIG_KPROBES_ON_FTRACE=y
CONFIG_HAVE_KPROBES=y
CONFIG_HAVE_KPROBES_ON_FTRACE=y
CONFIG_KPROBE_EVENTS=y
# CONFIG_KPROBES_SANITY_TEST is not set
```

Any suggestions?

**Comment by rafaeldtinoco** [comment]
@rainkin1993 are you setting the ebpf kernel version attribute ? It is mandatory for 4.15:

https://github.com/rafaeldtinoco/portablebpf/blob/9423966d077bd1cf15b904da76b2ae45349c8e41/mine.c#L232

and the result will be:

`libbpf: load bpf program failed: Invalid argument`

You can use [THIS](https://github.com/rafaeldtinoco/portablebpf/tree/devel) code as example and this is the [DELTA](https://github.com/rafaeldtinoco/portablebpf/blob/devel/patches/libbpf-introduce-legacy-kprobe-events-support.patch) to libbpf to make it work in v4.15.

> Make sure you install Ubuntu -dbgsym package for the kernel you are using AND, using a new pahole binary, do a `pahole -J /usr/lib/debug/boot/vmlinux-$(uname-r)`. After that, the BTFs will be included in the vmlinux file.

**Comment by anakryiko** [comment]
> I think that this approach here:
> 
> [rafaeldtinoco@e5be2dc#diff-ac82747bd450e8e5e0805a99d6c24afa9fcbfcddffbce5aa3288b2edabe17419L10302](https://github.com/rafaeldtinoco/libbpf/commit/e5be2dcbf4aedcc0a8d80b6eb0c74338bc93a243#diff-ac82747bd450e8e5e0805a99d6c24afa9fcbfcddffbce5aa3288b2edabe17419L10302)
> 
> is best as it only adds a single perf_event_open_probe_legacy() function to differentiate from what libbpf already has.

Yes, it looks pretty minimal as I hoped. I'll have few suggestions on how to structure things differently w.r.t. bpf_link, and I'm still not sure why skeleton destroy has to account for those legacy kprobes at all. But please submit what you have to the mailing list and we'll take it from there. Thanks!

**Comment by rainkin1993** [comment]
> @rainkin1993 are you setting the ebpf kernel version attribute ? It is mandatory for 4.15:
> 
> https://github.com/rafaeldtinoco/portablebpf/blob/9423966d077bd1cf15b904da76b2ae45349c8e41/mine.c#L232
> 
> and the result will be the following (if not being set):
> 
> `libbpf: load bpf program failed: Invalid argument`
> 
> > Ubuntu 4.15.0 kernel is actually based out of upstream 4.15.18, thus the version being .18.
> 
> You can use [THIS](https://github.com/rafaeldtinoco/portablebpf/tree/devel) code as example and this is the [DELTA](https://github.com/rafaeldtinoco/portablebpf/blob/devel/patches/libbpf-introduce-legacy-kprobe-events-support.patch) to libbpf to make it work in v4.15.
> 
> > Make sure you install Ubuntu -dbgsym package for the kernel you are using AND, using a new pahole binary, do a `pahole -J /usr/lib/debug/boot/vmlinux-$(uname-r)`. After that, the BTFs will be included in the vmlinux file.

@anakryiko 
Thanks. By following your answer, my code in kernel 4.18 also works.
The basic reason is that ubuntu kernel is a little different than mainline kernel.
like the above ubuntu 4.18.0-18-generic kernel is actually based on mainline 4.18.20 kernel.
I just use the following code to modify the ebpf kernel version field and then the code works well.
```
bpf_object__set_kversion(skel->obj, 0x41214);
// 0x41214 mean 4.18.20
```
BTW, the mapping between ubuntu kernel version and mainline kernel version can be founded in here: https://people.canonical.com/~kernel/info/kernel-version-map.html

**Comment by rafaeldtinoco** [comment]
> > I think that this approach here:
> > [rafaeldtinoco@e5be2dc#diff-ac82747bd450e8e5e0805a99d6c24afa9fcbfcddffbce5aa3288b2edabe17419L10302](https://github.com/rafaeldtinoco/libbpf/commit/e5be2dcbf4aedcc0a8d80b6eb0c74338bc93a243#diff-ac82747bd450e8e5e0805a99d6c24afa9fcbfcddffbce5aa3288b2edabe17419L10302)
> > is best as it only adds a single perf_event_open_probe_legacy() function to differentiate from what libbpf already has.
> 
> Yes, it looks pretty minimal as I hoped. I'll have few suggestions on how to structure things differently w.r.t. bpf_link, and I'm still not sure why skeleton destroy has to account for those legacy kprobes at all. But please submit what you have to the mailing list and we'll take it from there. Thanks!

Yes, sorry for the delay (again). I'm changing jobs (now will work with eBPF in a permanent basis =).  Problem here was that file descriptor needed to be closed before kprobe_events had the event removed:

```
 static int bpf_link__detach_perf_event(struct bpf_link *link)
 {
    int err;
@@ -10145,8 +10193,10 @@ static int bpf_link__detach_perf_event(struct bpf_link *link)
    err = ioctl(link->fd, PERF_EVENT_IOC_DISABLE, 0);
    if (err)
        err = -errno;
-
    close(link->fd);
+   if (link->legacy.name)
+       remove_kprobe_event_legacy(link->legacy.name, link->legacy.retprobe);
+
    return libbpf_err(err);
 }
```

**Comment by rafaeldtinoco** [comment]
Sending tomorrow. Thanks.

**Comment by rafaeldtinoco** [comment]
https://lore.kernel.org/bpf/20210625044459.1249282-1-rafaeldtinoco@gmail.com/

**Comment by rafaeldtinoco** [comment]
Im replying back to you this weekend (LKML thread). I believe next week with back and forth we can finish this.

**Comment by rafaeldtinoco** [comment]
Wow, I just saw that https://github.com/libbpf/libbpf/issues/318 is about what I just mentioned (I believe).

**Comment by rafaeldtinoco** [comment]
https://lore.kernel.org/bpf/20210801051123.3822498-1-rafaeldtinoco@gmail.com/

already to be applied on top of the:

[PATCH v3 bpf-next 00/14] BPF perf link and user-provided bpf_cookie

patch set.

Thanks for considering it.

**Comment by anakryiko** [comment]
Merged from upstream in #380 (https://github.com/libbpf/libbpf/pull/380/commits/fedb000c108b391e6e37eb62ba182caf7f6f6cf2)
